### ISSUE TITLE
SDK Interface function to override audio output programically

### ISSUE THREAD

**Issue by iospro** [jitsi/jitsi-meet#10878]
Dot menu has the option to change audio output (headphone vs dynamic).
<img width="398" alt="1" src="https://user-images.githubusercontent.com/1611869/151626044-9325fcd5-da45-4fbb-a821-6d42914aa578.png">
 
But it needs a tap by hand. There is no function in the SDK interface for this.

We try to make screen off by enabling proximity detector on iOS. It is possible despite SDK tries to override our proximity preset.
If we try to change audio mode from videoChat to any other to override sound route to to headphone SDK resets it to its defaults.

The only way to solve our task is:
1) to allow client to make proximity enabled manually
2) to give change sound route function in the interface

It seems to be very easy and useful thing

**Comment by saghul** [comment]
The SDK will enable the proximity sensor when in audio-only mode.

There is no API for neither of those, but I'd welcome a PR. We can discuss specifics if you get started.

**Comment by iospro** [comment]
Before start this (also now I have sudden problems with build SDK from GIT - seems to be some node version problems and misnames of classes) I try to repeat the logic of dot-menu option - change output route.
In audio only mode it solves the problem, it changes proximity behaviour also.

In the console we see 
[AudioMode] Selected device: Built-In Microphone

According to the react and native code it follows that we definitely call

RCT_EXPORT_METHOD(setAudioDevice:(NSString *)device
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject)

With device = "Built-In Microphone"


Now we try to reproduce this trick by calling private objc method

Class amClass = NSClassFromString(@"AudioMode");
id audio = [amClass new];
void(^resolve)(id) = ^(id v){};

[audio performSelector:@selector(setAudioDevice:resolve:reject:) withObject:@"Built-In Microphone" withObject:resolve];

It works for a moment, but after is changed by SDK, we see in the console
[AudioMode] Route changed, reapplying RTCAudioSession config

And speaker works again.


So we realize that SDK do something additional when we tap dot-menu output selection in SDK.


Could you explain in detail this process? 
If we succeed in this private way then we can try to make useful pull request.

**Comment by saghul** [comment]
Each device has a name, a type and a UID. setDevice needs to be called with the UID. That's what we do in JS: https://github.com/jitsi/jitsi-meet/blob/master/react/features/mobile/audio-mode/components/AudioRoutePickerDialog.js

**Comment by iospro** [comment]
We know exactly the same as it seems.

react code call the following on tap

    _onSelectDeviceFn(device: Device) {
        return () => {
            this._hide();
            AudioMode.setAudioDevice(device.uid || device.type);
        };
    }

here we can send device.uid or device.type.

In the Xcode console we see every time
`[AudioMode] Selected device: Built-In Microphone`

It is the UID of the build-in microphone. According to the AudioMode.m you set input audio device but not the output
```
RCT_EXPORT_METHOD(setAudioDevice:(NSString *)device
                  resolve:(RCTPromiseResolveBlock)resolve
                  reject:(RCTPromiseRejectBlock)reject) {
    DDLogInfo(@"[AudioMode] Selected device: %@", device);

...
        // Here we use AVAudioSession because RTCAudioSession doesn't expose availableInputs.
        AVAudioSession *_session = [AVAudioSession sharedInstance];
        AVAudioSessionPortDescription *port = nil;

        // Find the matching input device.
        for (AVAudioSessionPortDescription *portDesc in _session.availableInputs) {
            if ([portDesc.UID isEqualToString:device]) {
                port = portDesc;
                break;
            }
        }
```

**Comment by saghul** [comment]
Yes, we use the input because we'll use the same device for both input and output.

**Comment by iospro** [comment]
I managed to solve my audio problem by options configuration, without any private code.

But now I have another big problem: SDK doesn't work while screen is off by the proximity sensor.

When I try to start "audio mode only" call and cover proximity sensor then  SDK doesn't use any callbacks (for example "participantJoin"). When I release proximity sensor in this situation SDK fills the console log with participant join information. But I still have audio channel problems.

Also it is even impossible to override proximity behaviour.

To sum up I can use audio-only call only if I wait connection with uncovered proximity sensor.

**Comment by saghul** [comment]
That should be solved by this patch: https://github.com/jitsi/jitsi-meet/blob/master/patches/react-native%2B0.66.4.patch

If you build the SDK off master you should already have it applied.

**Comment by iospro** [comment]
Thanks, I will try to rebuild it.

My custom build SDK is from master but from December 2021.

**Comment by iospro** [comment]
If I set audio only and wait for another participant to come with proximity sensor covered there is no any conference.
It starts only after I uncover the sensor.

**Comment by saghul** [comment]
Do you have the proximity sensor patch applied? (It's in the patches/ directory in master, it should have been automatically applied when you run npm install.

**Comment by saghul** [comment]
Thanks for the report. We'll try to look into it.

**Comment by riyas145** [comment]
@saghul  when update this plugin for new version of RN?
